### PR TITLE
Added the ability to notify to email when the application stop

### PR DESCRIPTION
**Pull request by JPboane**
Refere to [this jira ticket](https://jira.fgh.org.mz/browse/EC-128)

### CODE REVIEW

File: camel-openmrs/src/main/java/org/openmrs/eip/component/camel/OpenmrsAttachmentProducer.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+import javax.activation.FileDataSource;
+
+import org.apache.camel.Exchange;
+import org.apache.camel.attachment.AttachmentMessage;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.context.ApplicationContext;
+
+public class OpenmrsAttachmentProducer extends AbstractOpenmrsProducer {
+	private static Logger log = LoggerFactory.getLogger(OpenmrsAttachmentProducer.class);
```
Comment (wluyima) [inline camel-openmrs/src/main/java/org/openmrs/eip/component/camel/OpenmrsAttachmentProducer.java:15]: Looks like you did not format your code

File: camel-openmrs/src/main/java/org/openmrs/eip/component/camel/OpenmrsAttachmentProducer.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+import javax.activation.DataHandler;
+import javax.activation.FileDataSource;
+
+import org.apache.camel.Exchange;
+import org.apache.camel.attachment.AttachmentMessage;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.context.ApplicationContext;
+
+public class OpenmrsAttachmentProducer extends AbstractOpenmrsProducer {
```
Comment (wluyima) [inline camel-openmrs/src/main/java/org/openmrs/eip/component/camel/OpenmrsAttachmentProducer.java:14]: Won't this break existing code that assumes only 2 producer subclass? Do you really have to extend AbstractOpenmrsProducer?

File: camel-openmrs/src/main/java/org/openmrs/eip/component/camel/OpenmrsAttachmentProducer.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+	
+	public OpenmrsAttachmentProducer(final OpenmrsEndpoint endpoint, final ApplicationContext applicationContext, final ProducerParams params) {
+		super(endpoint, applicationContext, params);
+	}
+	
+	@Override
+	public void process(final Exchange exchange) {
+		log.info("Attaching the log file to email");
+		
+		AttachmentMessage in = exchange.getMessage(AttachmentMessage.class);
```
Comment (wluyima) [inline camel-openmrs/src/main/java/org/openmrs/eip/component/camel/OpenmrsAttachmentProducer.java:25]: Do you really want to share the log file via email? The file might also be too big, I think the exception type that was thrown and stacktrace are enough

File: distribution/sender/routes/send-notification-email-route.xml [hunk 4]
```diff
@@ -0,0 +1,20 @@
+<routes xmlns="http://camel.apache.org/schema/spring">
```
Comment (wluyima) [inline distribution/sender/routes/send-notification-email-route.xml:1]: How different are these routes? isn't it possible to create a single route that is used by both the sender and receiver?

File: distribution/sender/routes/shutdown-route.xml [hunk 5]
```diff
@@ -6,8 +6,8 @@
             <method beanType="org.openmrs.eip.app.CustomFileOffsetBackingStore" method="disable()" />
         </script>
         
-        <!-- TODO Send a notification to the configured emails -->
-
+        <toD uri="direct:send-notification-email-route" />
```
Comment (wluyima) [inline distribution/sender/routes/shutdown-route.xml:9]: I think this needs to happen in a try finally block so that in case an error is encountered while sending the email, we are still able to shutdown the application

File: distribution/receiver/routes/shutdown-route.xml [hunk 6]
```diff
@@ -2,8 +2,8 @@
     <route id="shutdown-route">
         <from uri="direct:shutdown-route" />
 
-        <!-- TODO Send a notification to the configured emails -->
-
+      	<toD uri="direct:send-notification-email-route" />
```
Comment (wluyima) [inline distribution/receiver/routes/shutdown-route.xml:5]: I think this needs to happen in a try finally block so that in case an error is encountered while sending the email, we are still able to shutdown the application

File: camel-openmrs/src/main/java/org/openmrs/eip/component/camel/OpenmrsAttachmentProducer.java [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+	public void process(final Exchange exchange) {
+		log.info("Attaching the log file to email");
+		
+		AttachmentMessage in = exchange.getMessage(AttachmentMessage.class); 
+		
+		String logsURI = appContext.getEnvironment().getProperty("spring.email.attachment.application-log-file");
+		
+		File file = new File(logsURI);
+		
+		if (file.exists()) in.addAttachment("Dbsyn-logs", new DataHandler(new FileDataSource(logsURI)));
```
Comment (wluyima) [inline camel-openmrs/src/main/java/org/openmrs/eip/component/camel/OpenmrsAttachmentProducer.java:31]: Dbsyn**c**

File: camel-openmrs/src/main/java/org/openmrs/eip/component/camel/OpenmrsAttachmentProducer.java (hunk 2 above)
Comment (wluyima) [reply camel-openmrs/src/main/java/org/openmrs/eip/component/camel/OpenmrsAttachmentProducer.java:14]: You should avoid writing a custom component, endpoint or producer, please use a class implementing Processor and put your logic to the process method, this is the purpose of processors

File: camel-openmrs/src/main/java/org/openmrs/eip/component/camel/OpenmrsAttachmentProducer.java (hunk 1 above)
Comment (JPboane) [reply camel-openmrs/src/main/java/org/openmrs/eip/component/camel/OpenmrsAttachmentProducer.java:15]: Oooh, I see. I'll format it this time and next

File: camel-openmrs/src/main/java/org/openmrs/eip/component/camel/OpenmrsAttachmentProducer.java (hunk 2 above)
Comment (JPboane) [reply camel-openmrs/src/main/java/org/openmrs/eip/component/camel/OpenmrsAttachmentProducer.java:14]: Okay. I'll change it to Processor

File: camel-openmrs/src/main/java/org/openmrs/eip/component/camel/OpenmrsAttachmentProducer.java (hunk 3 above)
Comment (JPboane) [reply camel-openmrs/src/main/java/org/openmrs/eip/component/camel/OpenmrsAttachmentProducer.java:25]: Sometime you need the whole log to understand the context of the Exception. So to avoid to go back to site and ask for the log could be painful. The log file is setup to limit of 10mb. If this is too much we can compress it.

File: distribution/sender/routes/send-notification-email-route.xml (hunk 4 above)
Comment (JPboane) [reply distribution/sender/routes/send-notification-email-route.xml:1]: The only difference is that in the sender we know the senderId from the properties file. So the email cames with the sender name

File: distribution/sender/routes/shutdown-route.xml (hunk 5 above)
Comment (JPboane) [reply distribution/sender/routes/shutdown-route.xml:9]: Okay

File: distribution/receiver/routes/shutdown-route.xml (hunk 6 above)
Comment (JPboane) [reply distribution/receiver/routes/shutdown-route.xml:5]: Ok

File: distribution/sender/routes/send-notification-email-route.xml (hunk 4 above)
Comment (JPboane) [reply distribution/sender/routes/send-notification-email-route.xml:1]: And were should I put this shared route in case we dicide to use the same route? In the sender routes or in the receiver routes? Or should we create a new directory for shared routes?

File: camel-openmrs/src/main/java/org/openmrs/eip/component/camel/OpenmrsAttachmentProducer.java (hunk 3 above)
Comment (wluyima) [reply camel-openmrs/src/main/java/org/openmrs/eip/component/camel/OpenmrsAttachmentProducer.java:25]: OK, that makes sense especially from the context of the remote sites, and yes I guess you are right the log file size is limited

File: camel-openmrs/src/main/java/org/openmrs/eip/component/camel/OpenmrsAttachmentProducer.java (hunk 7 above)
Comment (wluyima) [reply camel-openmrs/src/main/java/org/openmrs/eip/component/camel/OpenmrsAttachmentProducer.java:31]: Did you see this comment?

File: distribution/receiver/routes/send-notification-email-route.xml [hunk 8]
```diff
@@ -0,0 +1,20 @@
+<routes xmlns="http://camel.apache.org/schema/spring">
+    <route id="send-notification-email-route">
```
[the remark this answers is not in the document]
Comment (wluyima) [reply distribution/receiver/routes/send-notification-email-route.xml:2]: Did you see this comment?

File: distribution/receiver/routes/send-notification-email-route.xml [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+       <!-- <from uri="scheduler:send-notification-email-route?initialDelay=1000&amp;delay=60000" /> -->
```
Comment (wluyima) [inline distribution/receiver/routes/send-notification-email-route.xml:3]: You can remove this line instead of commenting it out

File: distribution/sender/routes/send-notification-email-route.xml (hunk 4 above)
Comment (wluyima) [reply distribution/sender/routes/send-notification-email-route.xml:1]: A simple  approach is to move this logic to a processor instead of a route, the processor should be in the **camel-openmrs** sub module, then the processot can be called from both the sender and receiver

File: camel-openmrs/src/main/java/org/openmrs/eip/component/camel/OpenmrsAttachmentProducer.java (hunk 7 above)
Comment (JPboane) [reply camel-openmrs/src/main/java/org/openmrs/eip/component/camel/OpenmrsAttachmentProducer.java:31]: Yes

File: distribution/sender/routes/send-notification-email-route.xml (hunk 4 above)
Comment (JPboane) [reply distribution/sender/routes/send-notification-email-route.xml:1]: Okay. I'll do that

File: distribution/receiver/routes/send-notification-email-route.xml (hunk 8 above)
[the remark this answers is not in the document]
Comment (JPboane) [reply distribution/receiver/routes/send-notification-email-route.xml:2]: Yes

File: distribution/receiver/routes/send-notification-email-route.xml (hunk 9 above)
Comment (JPboane) [reply distribution/receiver/routes/send-notification-email-route.xml]: Ok

File: camel-openmrs/src/main/java/org/openmrs/eip/component/camel/SendShutdownNotificationEmailProcessor.java [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+	private String springEmailSendTo;
+	
+	@Value("${spring.profiles.active}")
+	private String activeProfile;
+	
+	@Value("${db-sync.senderId:#{null}}")
+	private String dbSyncSenderId;
+	
+	@Value("${db-sync.receiverId:#{null}}")
+	private String dbSyncReceiverId;
```
Comment (wluyima) [inline camel-openmrs/src/main/java/org/openmrs/eip/component/camel/SendShutdownNotificationEmailProcessor.java:49]: Move this property out of this processor to the receiver route where you call this processor class so that you don't have to make it nullable here

File: distribution/docs/sender/application.properties [hunk 11]
```diff
@@ -34,9 +34,6 @@ db-event.retry.interval=1800000
 
 # The interval in milliseconds before the first poll of the retry queue by the retry route when the application starts.
 db-event.retry.initial.delay=120000
-
-# Number of events to sync in parallel, defaults to 10
-events.sync.parallel.size=
 # ----------------------------------------------------------------------------------------------------------------------
```
Comment (wluyima) [inline distribution/docs/sender/application.properties:37]: Why are you removing this code?

File: distribution/docs/sender/application.properties [hunk 12]
```diff
@@ -232,3 +229,24 @@ pgp.sender.password=
 # User name of the PGP receiver key pair
 pgp.sender.receiverUserId=
 # ----------------------------------------------------------------------------------------------------------------------
+
+# ***************************** Application Status Notification: Emails Configuration ******************************************#
+#The list of email addresses to be notified when the application stops. Multiples addresses must be separated by comma
+spring.email.send-to=
+#Path to application log file
+spring.email.attachment.application-log-file=${eip.home}${file.separator}shared${file.separator}logs${file.separator}eip${file.separator}openmrs-eip.log
```
Comment (wluyima) [inline distribution/docs/sender/application.properties:237]: This is not the default log file location, it is **${eip.home}/logs/openmrs-eip.log**

File: distribution/receiver/routes/send-notification-email-route.xml (hunk 8 above)
[the remark this answers is not in the document]
Comment (wluyima) [reply distribution/receiver/routes/send-notification-email-route.xml:2]: Why do you need this route? Why don't you just out this logic in the shutdown-route for the receiver?

File: distribution/sender/routes/send-notification-email-route.xml [hunk 13]
```diff
@@ -0,0 +1,7 @@
...
+    <route id="send-notification-email-route">
```
Comment (wluyima) [inline distribution/sender/routes/send-notification-email-route.xml:2]: Move this logic to sender shutdown-route and get rid of this route

File: distribution/receiver/routes/shutdown-route.xml [hunk 14]
```diff
@@ -2,8 +2,14 @@
     <route id="shutdown-route">
         <from uri="direct:shutdown-route" />
 
-        <!-- TODO Send a notification to the configured emails -->
-
+      	<doTry>
+        	<toD uri="direct:send-notification-email-route" />
```
Comment (wluyima) [inline distribution/receiver/routes/shutdown-route.xml:6]: Call the processor directly from here and pass it the receiverId from here on the exchange

File: distribution/sender/routes/shutdown-route.xml [hunk 15]
```diff
@@ -6,8 +6,14 @@
             <method beanType="org.openmrs.eip.app.CustomFileOffsetBackingStore" method="disable()" />
         </script>
         
-        <!-- TODO Send a notification to the configured emails -->
-
+        <doTry>
+        	<toD uri="direct:send-notification-email-route" />
```
Comment (wluyima) [inline distribution/sender/routes/shutdown-route.xml:10]: Call the processor directly from here and pass it the sender from here on the exchange

File: distribution/receiver/routes/shutdown-route.xml [hunk 16]
```diff
@@ -2,8 +2,14 @@
     <route id="shutdown-route">
         <from uri="direct:shutdown-route" />
 
-        <!-- TODO Send a notification to the configured emails -->
-
+      	<doTry>
```
Comment (wluyima) [inline distribution/receiver/routes/shutdown-route.xml:5]: I'd think the try catch clause should be in the processor so that you don't duplicate the code in the sender and receiver

**Review by wluyima — COMMENTED** [review]
Add comments

**Comment by wluyima** [comment]
You have unresolved conflicts

File: camel-openmrs/src/main/java/org/openmrs/eip/component/camel/SyncActionEnum.java [hunk 17]
```diff
@@ -1,10 +1,10 @@
 package org.openmrs.eip.component.camel;
```
Comment (wluyima) [inline camel-openmrs/src/main/java/org/openmrs/eip/component/camel/SyncActionEnum.java:2]: Revert these changes

File: camel-openmrs/src/main/java/org/openmrs/eip/component/camel/SendShutdownNotificationEmailProcessor.java [hunk 18]
```diff
@@ -0,0 +1,84 @@
...
+		File file = new File(logsURI);
+		
+		if (file.exists()) {
+			DefaultAttachment att = new DefaultAttachment(new FileDataSource(logsURI));
+			att.addHeader("Content-Description", "Applicatio Logs file");
+			in.addAttachmentObject("Dbsync-logs", att);
+		}
+		
+		Endpoint endpoint = exchange.getContext().getEndpoint(
+		    "smtp://" + smtpHostName + ":" + smtpHostPort + "?username=" + smtpAuthUser + "&password=" + smtpAuthPass);
```
Comment (wluyima) [inline camel-openmrs/src/main/java/org/openmrs/eip/component/camel/SendShutdownNotificationEmailProcessor.java:77]: Shouldn't this be smtps?

**Comment by wluyima** [comment]
Addressed review comments and merged the code into master
